### DISCUSSION TITLE
Easy commit splitting

### DISCUSSION THREAD

**Discussion by ghost** [magit/magit#4783] [Ideas]
When selling magit, my pitch is always how it facilitates producing a "groomed" sequence of commits.  It goes without saying that I do quite a bit of grooming.

I use the [technique described](https://www.internalpointers.com/post/split-commit-into-smaller-ones-git) here a lot.

Unfortunately, it is a slightly clunky multi-step process.  It would be wonderful if magit offered a one-step command to split a specified commit.  It would perform steps 1 through 3 in the mentioned article (including the 'reset HEAD~1'), leaving me at

> Now all the changes done in that commit are unstaged and need to be committed again. This is the step where you create new smaller commits, or in other words where you split the original one. Commit the pieces individually in the usual way (git commit ...), producing as many commits as you need.

Once done I would simply 'r r'.  _Et, voilà_, my commit is split.

**Comment by kyleam** [comment]
I can see the appeal of a command like you propose, especially given that it looks like this is a bit more complicated than "if you want to do X then Y, write a light wrapper that chains X and Y" because the rebase is asynchronous.

Fwiw here's a lightly tested sketch:

```elisp
;;; -*- lexical-binding:t -*-

(defun my/magit-rebase-redo-commit (commit args)
  (interactive (list (magit-commit-at-point)
                     (magit-rebase-arguments)))
  (if commit
      (let ((base (concat commit "~")))
        (unless (magit-rev-verify base)
          (user-error "Commit %s has no parent" commit))
        (magit-rebase-edit-commit commit args)
        (set-process-sentinel
         magit-this-process
         (lambda (process event)
           (magit-sequencer-process-sentinel process event)
           (when (and (eq (process-status process) 'exit)
                      (= (process-exit-status process) 0))
             (magit-run-git "reset" base)))))
    (magit-log-select
      (lambda (commit)
        (my/magit-rebase-redo-commit commit args))
      "Type %p on a commit to redo,")))
```

**Reply by ghost** [reply]
Kyle,

Thank you for your quick suggestion.  I have been a bit busy or would have replied sooner.  Will try this out shortly.

/john

**Comment by glasserc** [comment]
I'd love to see a feature that made this easier, but I don't think we need to necessarily imitate the mechanisms we already use for splitting commits. I think the design space for splitting commits is still quite unexplored throughout the git ecosystem. I recently discovered [git-revise](https://github.com/mystor/git-revise) which offers a mechanism for splitting using an interface like `git add --interactive`. You choose the parts of the commit you wish were in another commit, and it encodes them as a new commit after your existing commit (which has the remaining changes). That sounds nicer to me than getting my hands dirty with `reset` and restaging all the changes all over again.

**Reply by twmr** [reply]
Thx a lot for mentioning git-revise @glasserc! I'll give it a try. Do you already use it from within emacs via magit?

**Reply by glasserc** [reply]
No -- I don't think it's possible to use `git-revise` from within `magit` yet. (`magit` knows how to stage files but I think `git-revise` does everything in-memory.) But I'd love to see something like that be developed!

**Comment by tarsius** [comment]
Closing in favor of #5563.
